basedtyping/reified_generic.py

```diff
@@ -0,0 +1,160 @@
+                f"Type {self._name} cannot be instantiated; "
+                f"use {self.__origin__.__name__}() instead"
+            )
+        self._check_generics_resolved()
+        result = cast(_ReifiedGenericMetaclass, self.__origin__)._actual_call(
+            *args, **kwargs
+        )
+        try:
+            result.__orig_class__ = self  # type:ignore[attr-defined]
+        except AttributeError:
+            pass
+        return result
+
+    def _check_generics_resolved(self) -> None:
+        if len(self.__parameters__) > 0:
+            raise UnboundTypeVarError(
+                f"generic alias with unbound generics detected: {self.__parameters__}"
+            )
+
+    def _type_vars(self) -> tuple[TypeVar, ...]:
+        """gets a ``tuple`` of all the ``TypeVar``s defined in the `__origin__`.
+
+        basically you should always use this instead of ``self.__parameters__``"""
+        return cast(tuple[TypeVar, ...], getattr(self.__origin__, "__parameters__"))
+
+    def _type_var_check(self, args: tuple[type, ...]) -> bool:
+        self._check_generics_resolved()
+        for parameter, self_arg, subclass_arg in zip(
+            self._type_vars(),
+            self.__args__,
+            args,
+            strict=True,
+        ):
+            if not parameter.__contravariant__:
+                # if it's covariant (or invariant in which case check both ways):
+                if not issubform(subclass_arg, self_arg):
+                    return False
+            if not parameter.__covariant__:
+                # if it's contravariant (or invariant in which case check both ways):
+                if not issubform(self_arg, subclass_arg):
+                    return False
+        return True
+
+    def __subclasscheck__(self, subclass: object) -> bool:
+        # could be any random class, check it first
+        if not isinstance(subclass, _ReifiedGenericAlias) or not issubform(
+            subclass.__origin__, self.__origin__
+        ):
+            return False
+        subclass._check_generics_resolved()
+        return self._type_var_check(subclass.__args__)
+
+    def __instancecheck__(self, instance: object) -> bool:
+        # could be any random instance, check it first
+        if not isinstance(instance, self.__origin__) or not isinstance(
+            instance, ReifiedGeneric
+        ):
+            return False
+        return self._type_var_check(instance.__orig_class__.__args__)
+
+
+class OrigClass(Protocol):
+    __args__: tuple[type, ...]
+    __parameters__: tuple[TypeVar, ...]
+
+
+class ReifiedGenericError(TypeError):
+    pass
+
+
+class NotReifiedError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated without using a generic alias (ie. the generics can't be reified)"""
+
+
+class UnboundTypeVarError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated, ``isinstance`` checked or ``issubclass`` checked with an unbound ``TypeVar``"""
```

Comment:
```suggestion
class NonReifiedTypeVarError(ReifiedGenericError):
    """Raised when a ``ReifiedGeneric`` is instanciated, ``isinstance`` checked or ``issubclass`` checked with a non reified ``TypeVar`` as a type parameter."""
```

Bound `TypeVar`s are still invalid as type parameters to a `ReifiedGeneric`.

Reply:
do u think i should delete one of these errors then? they are for different scenarios but mean the same thing?
https://github.com/KotlinIsland/basedtyping/pull/7/files#diff-7fd05312aa4b91c50fc6766fd2789845da2abc401733491608cca183e50ea6a8R112-R117

Reply:
One is for `Foo()`, one is for `Foo[T]()`.

If you think that's the same then merge. They do look a little diff to me

Reply:
theyre different but you're suggesting the term "not reified" to refer to both scenarios if i understand correctly? i was using "not reified" to refer to `Foo()` and "not bound" to refer to `Foo[T]()` which made sense to me. what do you suggest i rename these two errors to?